libs/coin-modules/coin-solana/src/synchronization.ts

```diff
@@ -81,79 +198,73 @@ export const getAccountShapeWithAPI = async (
   const subAccByMint = pipe(
     () => mainInitialAcc?.subAccounts ?? [],
-    filter((subAcc): subAcc is TokenAccount => subAcc.type === "TokenAccount"),
     keyBy(subAcc => subAcc.token.contractAddress),
     v => new Map(toPairs(v)),
   )();
 
-  const nextSubAccs: TokenAccount[] = [];
-
-  for (const [mint, accs] of onChainTokenAccsByMint.entries()) {
-    if (!tokenIsListedOnLedger(currency.id, mint)) {
-      continue;
-    }
-
-    const tokenProgram = accs[0].onChainAcc.account.data.program as SolanaTokenProgram;
-    const assocTokenAccAddress = await api.findAssocTokenAccAddress(
-      mainAccAddress,
-      mint,
-      tokenProgram,
-    );
-
-    const assocTokenAcc = accs.find(
-      ({ onChainAcc: { pubkey } }) => pubkey.toBase58() === assocTokenAccAddress,
-    );
-
-    if (assocTokenAcc === undefined) {
-      continue;
-    }
-
-    const subAcc = subAccByMint.get(mint);
-
-    const lastSyncedTxSignature = subAcc?.operations?.[0]?.hash;
-
-    const txs = await getTransactions(
-      assocTokenAcc.onChainAcc.pubkey.toBase58(),
-      lastSyncedTxSignature,
-      api,
-    );
-
-    const mintInfo =
-      tokenProgram === PARSED_PROGRAMS.SPL_TOKEN_2022
-        ? await getMaybeTokenMint(mint, api)
-        : undefined;
+  // all token accounts
+  const supportedOnChainTokenAccounts = onChainTokenAccounts.filter(({ info: { mint } }) =>
+    tokenIsListedOnLedger(currency.id, mint.toBase58()),
+  );
+  const supportedOnChainTokenAccountsByMint = groupBy(
+    supportedOnChainTokenAccounts,
+    account => account.info.mint,
+  );
+  const onChainAssociatedTokenAccounts = compact(
+    Object.entries<ParsedOnChainTokenAccountWithInfo[]>(supportedOnChainTokenAccountsByMint).map(
+      ([mint, accounts]) => toAssociatedTokenAccount(mint, mainAccAddress, accounts, subAccByMint),
+    ),
+  );
+  const onChainAssociatedTokenAccountsByAddress = Object.fromEntries(
+    onChainAssociatedTokenAccounts.map(account => [
+      account.associatedTokenAccount.onChainAcc.pubkey.toBase58(),
+      account,
+    ]),
+  );
 
-    const mintExtensions =
-      mintInfo && !(mintInfo instanceof Error) ? mintInfo.info.extensions : undefined;
+  const mints = Object.keys(supportedOnChainTokenAccountsByMint);
+  const mintInfos = (await api.getMultipleAccounts(mints))
+    .map(account =>
+      account && "parsed" in account.data ? tryParseAsMintAccount(account.data) : undefined,
+    )
+    .map(mintInfo => (mintInfo && !(mintInfo instanceof Error) ? mintInfo.extensions : undefined));
+  const mintInfosByMint = Object.fromEntries(mints.map((mint, i) => [mint, mintInfos[i]]));
 
-    const nextSubAcc =
-      subAcc === undefined
-        ? await newSubAcc({
+  const accountsWithTransactions = onChainAssociatedTokenAccounts.length
+    ? await getTokenAccountsTransactions(onChainAssociatedTokenAccounts, api)
+    : [];
+  const nextSubAccs: TokenAccount[] = await Promise.all(
+    accountsWithTransactions.map(account => {
```

Comment:
Looks like you're already doing the parallel calls we talked about today for the sub accounts
The only drawback maybe here is that we don't limit the number of concurrent promises